src/util/tmpl/fd_map_giant.c

```diff
@@ -0,0 +1,828 @@
+    return NULL;
+  }
+
+  FD_COMPILER_MFENCE();
+  FD_VOLATILE( hdr[0] ) = 0UL;
+  FD_COMPILER_MFENCE();
+
+  return shmap;
+}
+
+MAP_IMPL_STATIC MAP_T *
+MAP_(insert)( MAP_T *           join,
+              MAP_KEY_T const * key ) {
+  MAP_(private_t) * map = MAP_(private)( join );
+
+  /* Pop the free stack to allocate an element (this is guaranteed to
+     succeed as per contract) */
+
+  ulong ele_idx = map->free_stack;
+  MAP_T * ele = join + ele_idx;
+  map->free_stack = ele->MAP_NEXT & MAP_IDX_NULL;
+  map->key_cnt++; /* Consider eliminating this to help make completely concurrent lockfree? */
+
+  /* ... and map the newly allocated element to key (this is also
+     guaranteed to not have collisions as per contract). */
+
+  ulong * head = MAP_(private_list)( map ) + MAP_(private_list_idx)( key, map->seed, map->list_cnt );
+  MAP_(key_copy)( &ele->MAP_KEY, key );
+  ele->MAP_NEXT = *head;
+  *head = ele_idx;
+
+  return ele;
+}
+
+MAP_IMPL_STATIC MAP_T *
+MAP_(remove)( MAP_T *           join,
+              MAP_KEY_T const * key ) {
+  MAP_(private_t) * map = MAP_(private)( join );
+
+  ulong   list_cnt = map->list_cnt;
+  ulong * list     = MAP_(private_list)( map );
+
+  /* Find the key */
+
+  ulong * head = list + MAP_(private_list_idx)( key, map->seed, list_cnt );
+  ulong * cur  = head;
+  for(;;) {
+    ulong ele_idx = *cur;
+    if( FD_UNLIKELY( ele_idx==MAP_IDX_NULL ) ) break;
+    MAP_T * ele = join + ele_idx;
+    if( FD_LIKELY( MAP_(key_eq)( key, &ele->MAP_KEY ) ) ) { /* Optimize for found (it is remove after all) */
+      /* Found, remove the mapping and push it to free stack. */
+      *cur = ele->MAP_NEXT;
+      ele->MAP_NEXT = map->free_stack | (1UL<<63);
+      map->free_stack = ele_idx;
+      map->key_cnt--; 
+      return ele;
+    }
+    cur = &ele->MAP_NEXT; /* Retain the pointer to next so we can rewrite it later. */
+  }
+
+  /* Not found */
+
+  return NULL;
+}
+
+FD_FN_PURE MAP_IMPL_STATIC MAP_T *
+MAP_(query)( MAP_T *           join,
+             MAP_KEY_T const * key,
+             MAP_T *           null ) {
+  MAP_(private_t) * map = MAP_(private)( join );
+
+  ulong   list_cnt = map->list_cnt;
+  ulong * list     = MAP_(private_list)( map );
+
+  /* Find the key */
+
+  ulong * head = list + MAP_(private_list_idx)( key, map->seed, list_cnt );
+  ulong * cur  = head;
+  for(;;) {
+    ulong ele_idx = *cur;
+    if( FD_UNLIKELY( ele_idx==MAP_IDX_NULL ) ) break; /* optimize for found */
+    MAP_T * ele = join + ele_idx;
+    if( FD_LIKELY( MAP_(key_eq)( key, &ele->MAP_KEY ) ) ) { /* optimize for found */
+      /* Found, move it to the front of the chain. (FIXME: BRANCH PROB? DO BRANCHLESS?) */
+      if( cur!=head ) {
+        *cur = ele->MAP_NEXT;
+        ele->MAP_NEXT = *head;
+        *head = ele_idx;
+      }
+      return ele;
+    }
+    cur = &ele->MAP_NEXT; /* Retain the pointer to next so we can rewrite it later. */
+  }
+
+  /* Not found */
+
+  return null;
```

Comment:
Will rename to sentinel.  In practice, it is very useful to often give code the ability to pass in a default behavior so that they can suppress unnecessary branches in critical map lookups.

Reply:
Done, renamed sentinel.